### ISSUE TITLE
[Bug] "download_finished" event is fired before pyfile status is set

### ISSUE THREAD

**Issue by adrienyhuel** [pyload/pyload#2252]
In plugins/internal/Addon.py, the function called after each finished download is defined like that : 

``` python
#: Deprecated method, use `download_finished` instead (Remove in 0.4.10)
def downloadFinished(self, pyfile):
    if pyfile.hasStatus("finished"):  #: Check if still "finished" (Fix in 0.4.10)
        return self.download_finished(pyfile)
```

The check was added recently and it appears it doesn't work, I had to comment the line to execute the script

``` python
#: Deprecated method, use `download_finished` instead (Remove in 0.4.10)
def downloadFinished(self, pyfile):
    #if pyfile.hasStatus("finished"):  #: Check if still "finished" (Fix in 0.4.10)
        return self.download_finished(pyfile)
```

The problem is that in module/PluginThread.py, the call to hookManager.downloadFinished is in start of method "run": 

``` python
 try:
                ...
                # start download
                ...
                self.m.log.info(_("Download finished: %s") % pyfile.name)
                self.m.core.hookManager.downloadFinished(pyfile)
                ...
```

but the status of the "pyfile" is set at the end of the method (pyfile.finishIfDone()): 

``` python
finally:
            ...
            pyfile.finishIfDone()
            ...
```

So when it enters in Addon.py, the status of the file is not "finished" yet

**Comment by vuolter** [comment]
Nice shot!

**Comment by youlJunior** [comment]
@vuolter 
It was difficult to follow methods between classes, there are a lot of imports and overrides...

**Comment by vuolter** [comment]
Eh, I know, there are many code work-around too...
Anyway thank you for https://github.com/pyload/pyload/commit/97c403b9e8f8ef11c11f8cd5e33f7ffceb31e38c ;)

**Comment by youlJunior** [comment]
You're welcome. Like I already said I'm a python fan :-)
Anyway, I'm going to take a look at Git, in order to fork the project, and make pull requests ;-)

**Comment by vuolter** [comment]
:+1:

**Comment by Velociraptor85** [comment]
So @vuolter is this BUG closed? Or @youlJunior could you close the Issue?

**Comment by GammaC0de** [comment]
@Velociraptor85 This issue will be resolved in v0.4.10
